include/fi_mem.h

```diff
@@ -88,12 +101,65 @@ union util_buf {
 	uint8_t data[0];
 };
 
-struct util_buf_pool *util_buf_pool_create(size_t size, size_t alignment,
-					   size_t max_cnt, size_t chunk_cnt);
-void util_buf_pool_destroy(struct util_buf_pool *pool);
+/* create buffer pool with alloc/free handlers */
+struct util_buf_pool *util_buf_pool_create_ex(size_t size, size_t alignment,
+					    size_t max_cnt, size_t chunk_cnt,
+					    util_buf_region_alloc_hndlr alloc_hndlr,
+					    util_buf_region_free_hndlr free_hndlr);
+
+/* create buffer pool */
+static inline struct util_buf_pool *util_buf_pool_create(size_t size,
+							 size_t alignment,
+							 size_t max_cnt,
+							 size_t chunk_cnt)
+{
+	return util_buf_pool_create_ex(size, alignment, max_cnt, chunk_cnt, NULL, NULL);
+}
+
+static inline int util_buf_avail(struct util_buf_pool *pool)
+{
+	return !slist_empty(&pool->buf_list);
+}
 
 void *util_buf_get(struct util_buf_pool *pool);
+
+int util_buf_grow(struct util_buf_pool *pool);
+
+static inline void *util_buf_get_ex(struct util_buf_pool *pool, void **context)
+{
+	union util_buf *buf;
+	struct util_buf_footer *buf_ftr;
+
+	buf = util_buf_get(pool);
+	buf_ftr = (struct util_buf_footer *) ((char *) buf + pool->data_sz);
+	if (context)
+		*context = buf_ftr->region->context;
+	return buf;
+}
+
+static inline void *util_buf_alloc(struct util_buf_pool *pool)
+{
+	if (!util_buf_avail(pool)) {
+		if (util_buf_grow(pool))
+			return NULL;
+	}
+	return util_buf_get(pool);
+}
+
+static inline void *util_buf_alloc_ex(struct util_buf_pool *pool, void **context)
+{
+	union util_buf *buf;
+	struct util_buf_footer *buf_ftr;
+
+	buf = util_buf_alloc(pool);
+	buf_ftr = (struct util_buf_footer *) ((char *) buf + pool->data_sz);
+	if (context)
```

Comment:
maybe just assert(context) -- the app can call util_buf_alloc() directly if they don't need the context set.